### PR TITLE
Gkarpov/gpt2 rework

### PR DESCRIPTION
**Pull request by glkarpov**
Added GPT2MLP and GPT2Attention hierarchical blocks for the construction of GPT2 model.
Full test coverage.

### CODE REVIEW

File: wrappers/python/nntile/layer/gpt2_attention.py [hunk 1]
```diff
@@ -0,0 +1,982 @@
...
+                        redux=self.redux)
+        # dW_Q can be offloaded from GPU
+        self.w_q.grad.wont_use()
+        # X_Q can be offloaded from GPU
+        self.x_q.value.wont_use()
+        # dQ_transposed can be deleted
+        #self.q_transposed.grad.wont_use()
+        self.q_transposed.grad.invalidate_submit()
+
+    @staticmethod
```
Comment (Muxas) [inline wrappers/python/nntile/layer/gpt2_attention.py:807]: switch to
```
@classmethod
def from_torch (cls,...)
```
to change `__class__` into `cls`

File: wrappers/python/nntile/layer/gpt2_attention.py [hunk 2]
```diff
@@ -0,0 +1,982 @@
...
+        )
+
+        torch_layer.c_proj.weight.grad = torch.tensor(
+            to_numpy(self.w.grad)
+            .reshape(n_embd, n_embd).T
+        )
+        torch_layer.c_proj.bias.grad = torch.tensor(
+            to_numpy(self.out_proj_bias.grad)
+        )
+        return torch_layer
```
Comment (Muxas) [inline wrappers/python/nntile/layer/gpt2_attention.py:982]: no end-of-line symbol in the end of the file. Change settings of your VSCode to add EOL at the end of each file

File: wrappers/python/nntile/model/gpt2_mlp.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+        layers.append(new_layer)
+        activations.extend(new_layer.activations_output)
+        self.next_tag = next_tag
+        # Fill Base Model with the generated data
+        super().__init__(activations, layers)
+
+    # Randomly init all linear layers
+    def init_randn_async(self):
+        for layer in self.layers:
+            if type(layer) is Linear:
```
Comment (Muxas) [inline wrappers/python/nntile/model/gpt2_mlp.py:80]: such a function is present at `BaseModel`, but you limit it to `Linear`. Why do you even need it here? I believe that removing this function will not break testing code.

File: wrappers/python/tests/layer/test_gpt2_attention.py [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+def generate_inputs(dtype: str, params: GPT2AttentionTestParams):
+    rng = np.random.default_rng(42)
+    torch_layer_config = GPT2Config_torch(
+        n_embd=params.n_emb,
+        n_head=params.n_head,
+        use_cache=False,
+        resid_pdrop=0.0,
+        embd_pdrop = 0.0,
+        attn_pdrop=0.0,
+        scale_attn_weights = True
```
Comment (Muxas) [inline wrappers/python/tests/layer/test_gpt2_attention.py:88]: What does this parameter do?

File: wrappers/python/tests/layer/test_gpt2_attention.py [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+        nntile_layer.unregister()
+        nntile_layer.x_q.unregister()
+        nntile_layer.x_k.unregister()
+        nntile_layer.x_v.unregister()
+        nntile_layer.y.unregister()
+        print(torch.norm(y_nntile), torch.norm(y))
+        rtol = dtype2tol[dtype]['rtol']
+        assert torch.norm(y - y_nntile) <= rtol * torch.norm(y)
+
+    def test_forward_backward(self, starpu_simple, torch_rng, dtype: str,
```
Comment (Muxas) [inline wrappers/python/tests/layer/test_gpt2_attention.py:176]: rename to `test_backward`

File: wrappers/python/tests/layer/test_gpt2_attention.py [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+        nntile_layer.backward_async()
+        torch_layer_other = nntile_layer.to_torch_with_grads()
+        nntile_layer.unregister()
+        nntile_layer.x_q.unregister()
+        nntile_layer.x_k.unregister()
+        nntile_layer.x_v.unregister()
+        nntile_layer.y.unregister()
+
+        rtol = dtype2tol[dtype]['rtol']
+        assert torch.norm(y - y_nntile) <= rtol * torch.norm(y)
```
Comment (Muxas) [inline wrappers/python/tests/layer/test_gpt2_attention.py:193]: Instead of checking output, check gradient of inputs. This checking of output in `test_backward` was my mistake. I forgot to switch it to checking of gradient of inputs.

File: wrappers/python/tests/layer/test_gpt2_attention.py [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+    x_q_traits = TensorTraits(x_shape, x_basetile)
+    x_q_distr = [0] * x_q_traits.grid.nelems
+    x_value = x_type(x_q_traits, x_q_distr, 0)
+    x_grad = x_type(x_q_traits, x_q_distr, 0)
+    X = TensorMoments(x_value, x_grad, grad_required=True)
+
+    x_random = rng.standard_normal(x_shape)
+    x_nntile = np.array(x_random, dtype=np.float32, order="F")
+    x_value.from_array(x_nntile)
+    x_torch = torch.Tensor(x_nntile.T)
```
Comment (Muxas) [inline wrappers/python/tests/layer/test_gpt2_attention.py:120]: add `.requires_grad_(True)` as it will be needed to check gradient over input in an updated `test_backward`

File: wrappers/python/tests/model/test_gpt2_mlp.py [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+    x_distr = [0] * x_traits.grid.nelems
+    x_type = dtype2nntile[dtype]
+    x_value = x_type(x_traits, x_distr, 0)
+    x_grad = x_type(x_traits, x_distr, 0)
+    X = TensorMoments(x_value, x_grad, grad_required=True)
+    gen = np.random.default_rng(42)
+    x_random = gen.standard_normal(x_shape, dtype=np.float32)
+    x_nntile = np.array(x_random, dtype=np.float32, order="F")
+    x_value.from_array(x_nntile)
+    x_torch = torch.Tensor(x_nntile.T)
```
Comment (Muxas) [inline wrappers/python/tests/model/test_gpt2_mlp.py:117]: Add `.requires_grad_(True)` to help checking gradient over this input

File: wrappers/python/tests/model/test_gpt2_mlp.py [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+                     dtype: str):
+        torch_layer, nntile_layer, x, _ = generate_inputs(params, dtype)
+        y = torch_layer(x)
+        nntile_layer.forward_async()
+        y_nntile = torch.Tensor(to_numpy(nntile_layer.activations[-1].value).T)
+        nntile_layer.unregister()
+        rtol = dtype2tol[dtype]['rtol']
+        assert torch.norm(y - y_nntile) <= rtol * torch.norm(y)
+
+    def test_forward_backward(self, starpu_simple, torch_rng,
```
Comment (Muxas) [inline wrappers/python/tests/model/test_gpt2_mlp.py:162]: rename to `test_backward` and swap check of output by check of gradient over input

**Review by Muxas — CHANGES_REQUESTED** [review]

**Review by Muxas — APPROVED** [review]
